follower/execution/follower.go

```diff
@@ -0,0 +1,299 @@
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+const cacheSize = 1024
+
+type blockReader interface {
+	Notify() <-chan string
+	Read(filename string) ([]byte, error)
+}
+
+type Follower struct {
+	log zerolog.Logger
+
+	blocks blockReader
+	codec  dps.Codec
+	db     *badger.DB
+
+	data    map[uint64]*BlockData
+	current *BlockData
+	height  uint64
+	index   int
+
+	stop chan struct{}
+}
+
+func New(log zerolog.Logger, blocks blockReader, codec dps.Codec, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		blocks: blocks,
+		codec:  codec,
+		db:     db,
+
+		height: math.MaxUint64,
+		data:   make(map[uint64]*BlockData, cacheSize),
+
+		stop: make(chan struct{}),
+	}
+
+	return &f
+}
+
+// Run launches the execution follower.
+func (f *Follower) Run() error {
+	tick := time.NewTicker(1 * time.Second)
+	defer tick.Stop()
+
+	for range tick.C {
+		select {
+		case <-f.stop:
+			return nil
+		default:
+		}
+
+		// If cache is already full, wait for items to get consumed.
+		if len(f.data) == cacheSize {
+			continue
+		}
+
+		// If cache has available space, wait for new block data to be available.
+		select {
+		case <-f.stop:
+			return nil
+		case file := <-f.blocks.Notify():
+			b, err := f.blocks.Read(file)
+			if err != nil {
+				return fmt.Errorf("could not read block data: %w", err)
+			}
+
+			var data BlockData
+			err = f.codec.Unmarshal(b, &data)
+			if err != nil {
+				return fmt.Errorf("could not decode block data: %w", err)
+			}
+			if data.Block == nil {
+				return fmt.Errorf("invalid block data")
+			}
+
+			f.data[data.Block.Header.Height] = &data
+
+			// This should only ever happen once, when we get the oldest available block.
+			if data.Block.Header.Height < f.height {
+				if f.height != math.MaxUint64 {
+					// This should never happen, it would mean we received blocks in an
+					// incorrect order!
+					return fmt.Errorf("fatal discrepancy: received block data in incorrect order")
+				}
+				f.height = data.Block.Header.Height
+			}
+		}
+	}
+
+	return nil
+}
+
+// Update returns the next trie update, in chronological order. It is also in charge of moving onto the next
+func (f *Follower) Update() (*ledger.TrieUpdate, error) {
+	// If we reached the end of the trie updates for the current block, it means it should have been indexed
+	// successfully. Therefore, we move on to the next block.
+	if f.current == nil || f.index == len(f.current.TrieUpdates) {
+		err := f.next()
+		if err != nil {
+			return nil, fmt.Errorf("could not forward execution follower to the next block: %w", err)
+		}
+	}
+
+	// Copy next update to be returned.
+	update := f.current.TrieUpdates[f.index]
+	f.index++
+
+	return update, nil
+}
+
+func (f *Follower) next() error {
+	if len(f.data) == 0 {
+		return dps.ErrUnavailable
+	}
```

Comment:
It really doesn't make sense to use a sentinel error that is exported for a function call that is unexported.

Reply:
Why wouldn't it, if this error is then wrapped and returned by a function call that is exported? 🤔